Sources/ActorSingletonPlugin/ActorSingletonPlugin.swift

```diff
@@ -22,48 +23,70 @@ import DistributedActorsConcurrencyHelpers
 ///
 /// - Warning: Refer to the configured `AllocationStrategy` for trade-offs between safety and recovery latency for
 ///            the singleton allocation.
 /// - SeeAlso: The `ActorSingleton` mechanism is conceptually similar to Erlang/OTP's <a href="http://erlang.org/doc/design_principles/distributed_applications.html">`DistributedApplication`</a>,
 ///            and <a href="https://doc.akka.io/docs/akka/current/cluster-singleton.html">`ClusterSingleton` in Akka</a>.
-public final class ActorSingletonPlugin {
-    private var singletons: [String: BoxedActorSingleton] = [:]
-    private let singletonsLock = Lock()
-
-    public init() {}
-
-    // FIXME: document that may crash, it may right?
-    func ref<Message: Codable>(of type: Message.Type, settings: ActorSingletonSettings, system: ClusterSystem, props: _Props? = nil, _ behavior: _Behavior<Message>? = nil) throws -> _ActorRef<Message> {
-        try self.singletonsLock.withLock {
-            if let existing = self.singletons[settings.name] {
-                guard let proxy = existing.unsafeUnwrapAs(Message.self).proxy else {
-                    fatalError("Singleton [\(settings.name)] not yet initialized")
-                }
-                return proxy
-            }
-
-            let singleton = ActorSingleton<Message>(settings: settings, props: props, behavior)
-            try singleton.startAll(system)
-            self.singletons[settings.name] = BoxedActorSingleton(singleton)
+public actor ActorSingletonPlugin {
+    private var singletons: [String: (proxied: ActorID, proxy: any AnyActorSingletonProxy)] = [:]
+
+    public func proxy<Act>(
+        of type: Act.Type,
+        settings: ActorSingletonSettings,
+        system: ClusterSystem,
+        props: _Props = _Props(),
+        _ factory: ((ClusterSystem) async throws -> Act)? = nil
+    ) async throws -> Act
+        where Act: DistributedActor,
+        Act.ActorSystem == ClusterSystem
+    {
+        if let existingID = self.singletons[settings.name]?.proxied {
+            return try Act.resolve(id: existingID, using: system)
+        }
 
-            guard let proxy = singleton.proxy else {
-                fatalError("Singleton[\(settings.name)] not yet initialized")
-            }
+        // Spawn the proxy for the singleton (one per singleton per node)
+        let proxy = try await _Props.$forSpawn.withValue(_Props._wellKnownActor(name: "singletonProxy-\(settings.name)")) {
+            try await ActorSingletonProxy(
+                settings: settings,
+                system: system,
+                singletonProps: props,
+                factory
+            )
+        }
 
-            return proxy
+        // Assign the singleton actor a special id and set up the remote call interceptor
+        var id = _Props.$forSpawn.withValue(_Props._wellKnownActor(name: "singleton-\(settings.name)")) {
+            system.assignID(Act.self)
```

Comment:
TODO: note to self that this entire well known is not so great with DA and need to get away from this props via task local pattern here. 

I'll think about what we want to do here, but in general I think we should probably expose a new `assignID` to the system which takes various options how it assigns it... and this way we can use it from the singleton module, without these task-local trickery in here (in this specific instance we don't have to do it this way)

I'll follow up here with a patch to show what we could do so we can check it :)

Reply:
Followed up